Reduce number of copies by introducing space.posix_move

jrwdunham opened the pull request:
Connected to #314

@payten I've created a PR here just so we can discuss this work. I realize it's a WIP, but I thought this discussion might be useful as it progresses.

payten committed:
Introduce `posix_move` to allow POSIX filesystems to copy files directly from the source to the destination space

jrwdunham commented on storage_service/locations/api/resources.py:
```diff
@@ -381,19 +381,19 @@ def post_detail(self, request, *args, **kwargs):
                 destination_path = os.path.join(
                     destination_location.relative_path, destination_path)
-                origin_space.move_to_storage_service(
+                origin_space.move_direct(
```
Should we be trying `move_direct` and falling back to the old way if a `NotImplementedError` is raised?

payten replied:
Thanks @jrwdunham, appreciate the review.

Great idea, thanks.

jrwdunham commented on storage_service/locations/models/space.py:
```diff
@@ -261,6 +261,28 @@ def delete_path(self, delete_path, *args, **kwargs):
+        LOGGER.debug('move_direct: absolute destination_path: %s', destination_path)
+
+        try:
+            self.get_child_space().move_direct(
```
`Space.move_direct` needs to return the return value of `self.get_child_space().move_direct()` and not just call it; this is required by `storage_effects = v.src_space.move_direct` in models/package.py where, as things stand, `storage_effects` will be `None`.

payten replied:
Ah thanks! We'll take a closer look at the other usages/implementers of the move_to_storage_service/move_from_storage_service to make we haven't missed anything else like that.

payten commented:
Hi @jrwdunham, @jhsimpson,

We've cleaned up and refocused the code so it's clear we're only supporting the new "direct move" behaviour on POSIX filesystems.  As discussed offline, making this code work for different protocols could be done but would require further refactoring.  One possible approach would be to extend the space API so a source space produces a generic bytestream which the destination space could then consume. However, we believe the current approach should solve the immediate needs.

~I'll note that there's a good chance this code will conflict with the refactored code in the `async` branch.  Let us know if you'd like this code rebased at all and we'll make sure it's all still worky.~ `async` has been merged and we've rebased this branch with the latest code.

Thanks!

jrwdunham commented on storage_service/locations/models/space.py:
```diff
@@ -648,6 +669,19 @@ def _delete_path_local(self, delete_path):
+            hasattr(source_space.get_child_space(), 'posix_move') and
+            hasattr(destination_space.get_child_space(), 'POSIX_FILESYSTEM') and
+            destination_space.get_child_space().POSIX_FILESYSTEM and
+            hasattr(destination_space.get_child_space(), 'posix_move'))
```
Is it not true that a space-type model will have `POSIX_FILESYSTEM = True` if and only if it also has a `posix_move` method defined? If so, then I think you can remove the `POSIX_FILESYSTEM` class attribute altogether and simplify this test to just `return (hasattr(source_space.get_child_space(), 'posix_move') and hasattr(destination_space.get_child_space(), 'posix_move')`.

If what I said is not true, then two of these conditions could at least be removed by using `getattr`: `getattr(source_space.get_child_space(), 'POSIX_FILESYSTEM', False)`.

jrwdunham commented on storage_service/locations/models/local_filesystem.py:
```diff
@@ -51,3 +55,8 @@ def verify(self):
         self.space.verified = verified
         self.space.last_verified = datetime.datetime.now()
+
+    def posix_move(self, source_path, destination_path, destination_space, package=None, *args, **kwargs):
```
You don't need `*args, **kwargs` here, right? I mean all of the `posix_move` methods have the same signature. I think `*args, **kwargs` should be removed from the definitions of `posix_move` in local_filesystem.py, nfs.py, and space.py.

jrwdunham commented on storage_service/locations/api/resources.py:
```diff
@@ -346,19 +346,27 @@ def _move_files_between_locations(self, files, origin_location, destination_loca
+                )
+                origin_space.post_move_to_storage_service()
+                destination_space.move_from_storage_service(
+                    source_path=destination_path,
```
The required `destination_path` arg is missing here.

jrwdunham commented on storage_service/locations/models/space.py:
```diff
@@ -261,6 +261,27 @@ def delete_path(self, delete_path, *args, **kwargs):
+        LOGGER.debug('posix_move: destination_space.path: %s', destination_space.path)
+
+        if not Space.posix_move_supported(self, destination_space):
+            raise ValidationError(_('both spaces must be POSIX filesystems'))
```
You can just call the static method on the instance, i.e., `self.posix_move_supported`. It seems odd to call it from the `Space` class here.

Also, the Django `ValidationError` seems odd here to me given that this is not really form validation. Probably better to just use Python's builtin `TypeError`.

jrwdunham replied:
Also, it seems odd, to duplicate the call to `posix_move_supported` once before you call `posix_move` and once inside it. Why not just attempt to call `.posix_move` and accept the `TypeError` and fallback to `move_to_storage_service`/`move_from_storage_service`, e.g., in resources.py? I can see that the hairiness in package.py may make such an "ask for forgiveness" approach hard (impossible?). If it could be done, I think it would be preferable.

jrwdunham commented on storage_service/locations/models/package.py:
```diff
@@ -674,49 +674,81 @@ def _store_aip_to_pending(self, origin_location, origin_path):
-        storage_effects = v.dest_space.move_from_storage_service(
-            source_path=self.current_path,  # This should include Location.path
-            destination_path=os.path.join(
+        if Space.posix_move_supported(v.src_space, v.dest_space):
```
As detailed more below, I wonder if you could not instead refactor this to try the `posix_move` call and fallback to the `move_to/from` call if a `TypeError` is raised, which would remove the double call to `posix_move_supported`.

jrwdunham commented on storage_service/locations/models/package.py:
```diff
@@ -674,49 +674,81 @@ def _store_aip_to_pending(self, origin_location, origin_path):
+            # Both spaces are POSIX filesystems and support `posix_move`
+            #
+            # 1. move direct to the SS destination space/location,
+            # 2. setting the status to "uploaded" (if applicable),
```
The "if applicable" parenthetical doesn't really apply here if the LOM/ARKIVUM condition has been removed.

jrwdunham reviewed (commented):
@payten I haven't actually tested this code by running transfers through AM yet, but I have some comments. (I can't approve or request changes, I think because the PR was technically opened by me.) I'll do some actual testing and may add more comments after.

payten committed:
Fix up posix_move and related changes from code review:
- remove POSIX_FILESYSTEM flag and use presence of posix_move as the trigger for this POSIX move direct behaviour
- remove superfluous arg/kwargs params
- reinstate dropped destination_path
- introduce custom exception to control when posix_move fallback is required
- reword posix_move step description

payten commented:
Thanks for the help @jrwdunham! We've cleaned up those bits and fixed the dropped argument - thanks for catching that.

To avoid ambiguity, we thought it might be nice to have a custom exception, `PosixMoveUnsupportedError`, rather than using `TypeError`. Does that seem ok?  And changing it to use an exception was ok too.

Also, happy to close this out and create a new PR if that would be easier for the review process? And will squash all commits if preferred.

jrwdunham commented:
@payten thanks. This is looking good to me. I tested with it and confirmed that LocalFilesystem-type space (LFS) to LFS works: `posix_move` gets called and doesn't raise an exception. I also changed the name of that method in `LocalFilesystem` in my test deploy and confirmed that that causes it to fallback to the old spoke->hub->spoke-style move.

Adding the new commit was helpful for me during code review, but yeah, I think squashing these would be good in this case. I don't think a new PR is necessary.

sevein commented:
I've realized that the squashed merge resulted in showing the author as @jrwdunham instead of @payten. Sorry about that and props to @payten!
